streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java

```diff
@@ -64,9 +65,14 @@ public StoreChangelogReader(final Consumer<byte[], byte[]> restoreConsumer,
 
     @Override
     public void register(final StateRestorer restorer) {
-        restorer.setUserRestoreListener(userStateRestoreListener);
-        stateRestorers.put(restorer.partition(), restorer);
-        needsInitializing.put(restorer.partition(), restorer);
+        if (!stateRestorers.containsKey(restorer.partition())) {
```

Comment:
Okay this is a bit hack but is the key of the fix: I've decided to keep the restorer in the map and note replacing it with the new restorer, since from the ProcessorStateManager the checkpoint offset cannot be inferred. This is just to walk around the code hierarchy of Restorer and the ProcessorStateManager.